File: tests/Unit/PointwiseFunctions/Hydro/Python/Test_StressEnergy.py
# Distributed under the MIT License.
# See LICENSE.txt for details.

import numpy as np


def inverse_spacetime_metric(lapse, shift, inverse_spatial_metric):
    result = np.zeros([4, 4])
    result[0, 0] = -1.0 / (lapse * lapse)
    for i in range(3):
        result[0, i + 1] = -shift[i] * result[0, 0]
        result[i + 1, 0] = result[0, i + 1]
    for i in range(3):
        for j in range(3):
            result[i + 1, j + 1] = inverse_spatial_metric[i, j] + (
                shift[i] * shift[j] * result[0, 0]
            )
    return result


def four_velocity(spatial_velocity, shift, lorentz_factor, lapse):
    result = np.zeros(4)
    result[0] = lorentz_factor / lapse
    result[1:] = lorentz_factor * (spatial_velocity - shift / lapse)
    return result


def comoving_magnetic_field(
    spatial_velocity,
    magnetic_field,
    magnetic_field_dot_spatial_velocity,
    lorentz_factor,
    shift,
    lapse,
):
    result = np.zeros(4)
    result[0] = magnetic_field_dot_spatial_velocity / lapse
    result[1:] = (
        magnetic_field
        + lapse
        * result[0]
        * (lorentz_factor * (spatial_velocity - shift / lapse))
    ) / lorentz_factor
    return result


def energy_density(
    rest_mass_density,
    specific_enthalpy,
    pressure,
    lorentz_factor,
    magnetic_field_dot_spatial_velocity,
    comoving_magnetic_field_squared,
):
    return (
        rest_mass_density * specific_enthalpy * lorentz_factor**2
        - pressure
        + comoving_magnetic_field_squared * (lorentz_factor**2 - 0.5)
        - (lorentz_factor * magnetic_field_dot_spatial_velocity) ** 2
    )


def momentum_density(
    rest_mass_density,
    specific_enthalpy,
    spatial_velocity,
    lorentz_factor,
    magnetic_field,
    magnetic_field_dot_spatial_velocity,
    comoving_magnetic_field_squared,
):
    return (
        rest_mass_density
        * specific_enthalpy
        * lorentz_factor**2
        * spatial_velocity
        + comoving_magnetic_field_squared
        * lorentz_factor**2
        * spatial_velocity
        - magnetic_field_dot_spatial_velocity * magnetic_field
        - magnetic_field_dot_spatial_velocity**2
        * lorentz_factor**2
        * spatial_velocity
    )


def stress_trace(
    rest_mass_density,
    specific_enthalpy,
    pressure,
    spatial_velocity_squared,
    lorentz_factor,
    magnetic_field_dot_spatial_velocity,
    comoving_magnetic_field_squared,
):
    return (
        3.0 * pressure
        + rest_mass_density * specific_enthalpy * (lorentz_factor**2 - 1.0)
        + comoving_magnetic_field_squared
        * (lorentz_factor**2 * spatial_velocity_squared + 0.5)
        - magnetic_field_dot_spatial_velocity**2
        * (lorentz_factor**2 * spatial_velocity_squared + 1.0)
    )


def stress_energy_tensor(
    rest_mass_density,
    specific_internal_energy,
    pressure,
    lorentz_factor,
    lapse,
    comoving_magnetic_field_magnitude,
    spatial_velocity,
    shift,
    magnetic_field,
    spatial_metric,
    inverse_spatial_metric,
):
    inverse_spacetime_metric_v = inverse_spacetime_metric(
        lapse, shift, inverse_spatial_metric
    )

    magnetic_field_dot_spatial_velocity = np.dot(
        np.dot(spatial_metric, magnetic_field), spatial_velocity
    )
    comoving_magnetic_field_v = comoving_magnetic_field(
        spatial_velocity,
        magnetic_field,
        magnetic_field_dot_spatial_velocity,
        lorentz_factor,
        shift,
        lapse,
    )
    four_velocity_v = four_velocity(
        spatial_velocity, shift, lorentz_factor, lapse
    )
    rho_h_star = (
        (rest_mass_density + rest_mass_density * specific_internal_energy)
        + pressure
        + comoving_magnetic_field_magnitude**2
    )
    p_star = pressure + 0.5 * (comoving_magnetic_field_magnitude**2)

    result = np.zeros([4, 4])
    result[0, 0] = (
        (rho_h_star * (four_velocity_v[0] ** 2))
        + (p_star * inverse_spacetime_metric_v[0, 0])
        - (comoving_magnetic_field_v[0] ** 2)
    )
    result[1, 0] = (
        (rho_h_star * four_velocity_v[0] * four_velocity_v[1])
        + (p_star * inverse_spacetime_metric_v[0, 1])
        - (comoving_magnetic_field_v[0] * comoving_magnetic_field_v[1])
    )
    result[1, 1] = (
        (rho_h_star * four_velocity_v[1] ** 2)
        + (p_star * inverse_spacetime_metric_v[1, 1])
        - (comoving_magnetic_field_v[1] ** 2)
    )
    result[0, 1] = result[1, 0]
    result[2, 0] = (
        (rho_h_star * four_velocity_v[2] * four_velocity_v[0])
        + (p_star * inverse_spacetime_metric_v[2, 0])
        - (comoving_magnetic_field_v[2] * comoving_magnetic_field_v[0])
    )
    result[0, 2] = result[2, 0]
    result[2, 1] = (
        (rho_h_star * four_velocity_v[2] * four_velocity_v[1])
        + (p_star * inverse_spacetime_metric_v[2, 1])
        - (comoving_magnetic_field_v[2] * comoving_magnetic_field_v[1])
    )
    result[1, 2] = result[2, 1]
    result[2, 2] = (
        (rho_h_star * four_velocity_v[2] ** 2)
        + (p_star * inverse_spacetime_metric_v[2, 2])
        - (comoving_magnetic_field_v[2] ** 2)
    )
    result[3, 0] = (
        (rho_h_star * four_velocity_v[3] * four_velocity_v[0])
        + (p_star * inverse_spacetime_metric_v[3, 0])
        - (comoving_magnetic_field_v[3] * comoving_magnetic_field_v[0])
    )
    result[0, 3] = result[3, 0]
    result[3, 1] = (
        (rho_h_star * four_velocity_v[3] * four_velocity_v[1])
        + (p_star * inverse_spacetime_metric_v[3, 1])
        - (comoving_magnetic_field_v[3] * comoving_magnetic_field_v[1])
    )
    result[1, 3] = result[3, 1]
    result[3, 2] = (
        (rho_h_star * four_velocity_v[3] * four_velocity_v[2])
        + (p_star * inverse_spacetime_metric_v[3, 2])
        - (comoving_magnetic_field_v[3] * comoving_magnetic_field_v[2])
    )
    result[2, 3] = result[3, 2]
    result[3, 3] = (
        (rho_h_star * four_velocity_v[3] ** 2)
        + (p_star * inverse_spacetime_metric_v[3, 3])
        - (comoving_magnetic_field_v[3] ** 2)
    )
    return result

File: tests/Unit/PointwiseFunctions/Hydro/Test_StressEnergy.cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#include "Framework/TestingFramework.hpp"

#include <cstddef>
#include <string>

#include "DataStructures/DataBox/DataBox.hpp"
#include "DataStructures/DataVector.hpp"
#include "DataStructures/Tensor/Tensor.hpp"
#include "Framework/CheckWithRandomValues.hpp"
#include "Framework/SetupLocalPythonEnvironment.hpp"
#include "Helpers/DataStructures/DataBox/TestHelpers.hpp"
#include "PointwiseFunctions/Hydro/StressEnergy.hpp"
#include "PointwiseFunctions/Hydro/Tags.hpp"

namespace hydro {

SPECTRE_TEST_CASE("Unit.PointwiseFunctions.Hydro.StressEnergy",
                  "[Unit][Hydro]") {
  pypp::SetupLocalPythonEnvironment local_python_env(
      "PointwiseFunctions/Hydro/Python/");
  const DataVector used_for_size(5);
  pypp::check_with_random_values<1>(&energy_density<DataVector>,
                                    "Test_StressEnergy", {"energy_density"},
                                    {{{0.0, 1.0}}}, used_for_size);
  pypp::check_with_random_values<1>(&momentum_density<DataVector>,
                                    "Test_StressEnergy", {"momentum_density"},
                                    {{{0.0, 1.0}}}, used_for_size);
  pypp::check_with_random_values<1>(&stress_trace<DataVector>,
                                    "Test_StressEnergy", {"stress_trace"},
                                    {{{0.0, 1.0}}}, used_for_size);
  pypp::check_with_random_values<1>(
      &stress_energy_tensor<DataVector>, "Test_StressEnergy",
      {"stress_energy_tensor"}, {{{0.0, 1.0}}}, used_for_size);

  // Check compute item works correctly in DataBox
  // For simplicity, just do flat spacetime, spatial_velocity = (0.6, 0.0, 0.0),
  // magnetic_field = (0.5, 0.0, 0.0).
  TestHelpers::db::test_compute_tag<
      Tags::StressEnergyTensorCompute<DataVector>>("StressEnergyTensor");

  const Scalar<DataVector> rest_mass_density{{{DataVector{5, 0.45}}}};
  const Scalar<DataVector> specific_internal_energy{{{DataVector{5, 0.34}}}};
  const Scalar<DataVector> pressure{{{DataVector{5, 0.23}}}};
  const Scalar<DataVector> lorentz_factor{{{DataVector{5, 1.5625}}}};
  const Scalar<DataVector> comoving_magnetic_field_magnitude{
      {{DataVector{5, 0.5660388679}}}};
  const Scalar<DataVector> lapse{{{DataVector{5, 1.}}}};
  auto spatial_velocity =
      make_with_value<tnsr::I<DataVector, 3>>(rest_mass_density, 0.);
  const auto shift =
      make_with_value<tnsr::I<DataVector, 3>>(rest_mass_density, 0.);
  auto magnetic_field =
      make_with_value<tnsr::I<DataVector, 3>>(rest_mass_density, 0.);
  auto spatial_metric =
      make_with_value<tnsr::ii<DataVector, 3>>(rest_mass_density, 0.);
  auto inverse_spatial_metric =
      make_with_value<tnsr::II<DataVector, 3>>(rest_mass_density, 0.);

  tnsr::AA<DataVector, 3> result{};

  spatial_velocity.get(0) = 0.6;
  magnetic_field.get(0) = 0.5;
  for (size_t i = 0; i < 3; ++i) {
    spatial_metric.get(i, i) = 1.0;
    inverse_spatial_metric.get(i, i) = 1.0;
  }

  stress_energy_tensor(make_not_null(&result), rest_mass_density,
                       specific_internal_energy, pressure, lorentz_factor,
                       comoving_magnetic_field_magnitude, lapse,
                       spatial_velocity, shift, magnetic_field, spatial_metric,
                       inverse_spatial_metric);

  const auto box = db::create<
      db::AddSimpleTags<
          Tags::RestMassDensity<DataVector>,
          Tags::SpecificInternalEnergy<DataVector>, Tags::Pressure<DataVector>,
          Tags::LorentzFactor<DataVector>, ::gr::Tags::Lapse<DataVector>,
          Tags::ComovingMagneticFieldMagnitude<DataVector>,
          Tags::SpatialVelocity<DataVector, 3, Frame::Inertial>,
          ::gr::Tags::Shift<DataVector, 3, Frame::Inertial>,
          Tags::MagneticField<DataVector, 3, Frame::Inertial>,
          ::gr::Tags::SpatialMetric<DataVector, 3, Frame::Inertial>,
          ::gr::Tags::InverseSpatialMetric<DataVector, 3, Frame::Inertial>>,
      db::AddComputeTags<Tags::StressEnergyTensorCompute<DataVector>>>(
      rest_mass_density, specific_internal_energy, pressure, lorentz_factor,
      comoving_magnetic_field_magnitude, lapse, spatial_velocity, shift,
      magnetic_field, spatial_metric, inverse_spatial_metric);
  CHECK(db::get<Tags::StressEnergyTensorCompute<DataVector>>(box) == result);
}
}  // namespace hydro
